planar/coord/utm/utm.go

```diff
@@ -0,0 +1,484 @@
+/*
+Package utm provides the ability to work with UTM grids
+
+Ref: https://stevedutch.net/FieldMethods/UTMSystem.htm
+Ref: https://gisgeography.com/central-meridian/
+
+*/
+package utm
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/go-spatial/geom/planar/coord"
+)
+
+const (
+	k0        = 0.9996 // k0 - 0.9996  for UTM
+	l0        = 0      // λ0 = center of the map. == 0 for us.
+	precision = 0.0001
+)
+
+// UTM Zone Letters
+const (
+	ZoneC ZoneLetter = 'C'
+	ZoneD ZoneLetter = 'D'
+	ZoneE ZoneLetter = 'E'
+	ZoneF ZoneLetter = 'F'
+	ZoneG ZoneLetter = 'G'
+	ZoneH ZoneLetter = 'H'
+	ZoneI ZoneLetter = 'I'
+	ZoneJ ZoneLetter = 'J'
+	ZoneK ZoneLetter = 'K'
+	ZoneL ZoneLetter = 'L'
+	ZoneM ZoneLetter = 'M'
+	ZoneN ZoneLetter = 'N'
+	ZoneP ZoneLetter = 'P'
+	ZoneQ ZoneLetter = 'Q'
+	ZoneR ZoneLetter = 'R'
+	ZoneS ZoneLetter = 'S'
+	ZoneT ZoneLetter = 'T'
+	ZoneU ZoneLetter = 'U'
+	ZoneV ZoneLetter = 'V'
+	ZoneW ZoneLetter = 'W'
+	ZoneX ZoneLetter = 'X'
+)
+
+// ZoneLetter describes the UTM zone letter
+type ZoneLetter byte
+
+// String implements the stringer interface
+func (zl ZoneLetter) String() string { return fmt.Sprintf("%c", zl) }
+
+// IsNorthern returns if the Zone is in the northern hemisphere
+func (zl ZoneLetter) IsNorthern() bool { return zl >= 'N' }
+
+// IsValid will run validity check on the zone letter and number
+func (zl ZoneLetter) IsValid() bool { return zl >= 'C' && zl <= 'X' && zl != 'O' }
+
+// quick lookup table for central meridian for each zone
+// this can be calculated using the formula i :=  zone > 30 ? zone-31 :  30 - zone
+// cmd := 3 + (6*i)
+var centralMeridianDegrees = []uint{
+	3, 9, 15, 21, 27, 33, 39, 45, 51, 57, 63, 69, 75, 81, 87, 93, 99, 105, 111, 117, 123, 129, 135, 141, 147, 153, 159, 165, 171, 177,
+}
+
+func CentralMeridian(zone Zone) (degree int, err error) {
+
+	if !zone.IsValid() {
+		return 0, fmt.Errorf("zone is invalid")
+	}
+
+	if zone.Number <= 30 {
+		idx := 30 - zone.Number
+		return -1 * int(centralMeridianDegrees[idx]), nil
+	}
+	idx := zone.Number - 31
+	return int(centralMeridianDegrees[idx]), nil
+
+}
+
+var digraphZones = [...][2][4]rune{
+	[2][4]rune{
+		[4]rune{'V', 'U', 'T', 'S'},
+		[4]rune{'W', 'X', 'Y', 'X'},
+	},
+	[2][4]rune{
+		[4]rune{'D', 'C', 'B', 'A'},
+		[4]rune{'E', 'F', 'G', 'H'},
+	},
+	[2][4]rune{
+		[4]rune{'M', 'L', 'K', 'J'},
+		[4]rune{'N', 'P', 'Q', 'R'},
+	},
+}
+
+var latDigraphZones = [...]rune{'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'J', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'A', 'B', 'C', 'D', 'E'}
+
+type Digraph [2]rune
+
+func (d Digraph) String() string { return string(d[:]) }
+
+func newDigraph(zone Zone, lnglat coord.LngLat) (Digraph, error) {
+	if !zone.IsValid() {
+		return Digraph{}, fmt.Errorf("zone is invalid")
+	}
+	// Howmany times have we cycled through the zones.
+	dZone := digraphZones[zone.Number%3]
+	// We already know the zone is valid
+	cm, _ := CentralMeridian(zone)
```

Comment:
is the only error that can be reported an invalid zone? If so, why the extra check above and ignore the error here rather than just return the error here?

Reply:
Yes, the only error is the Zone is invalid. We need the check to make sure the zone is valid for the dZone line above this.